umbra-js/test/Umbra.test.ts

```diff
@@ -176,172 +206,248 @@ describe('Umbra class', () => {
-      verifyEqualValues(await dai.balanceOf(sponsorWallet.address), sponsorFee);
+  const sendTests = [
+    { id: 'send', name: 'Send' },
+    { id: 'batchSend', name: 'Batch send' },
+  ];
+
+  for (const test of sendTests) {
+    describe(`${test.name}, scan, and withdraw funds`, () => {
+      beforeEach(() => {
+        // Seems we somehow lose the provider attached to our sender, so make sure it's there. Without this
+        // some tests below throw with "Error: missing provider (operation="sendTransaction", code=UNSUPPORTED_OPERATION, version=abstract-signer/5.0.12)"
+        sender = sender.connect(ethers.provider);
+      });
+
+      const mintAndApproveDai = async (signer: Wallet, user: string, amount: BigNumber) => {
+        await dai.connect(signer).mint(user, amount);
+        await dai.connect(signer).approve(umbra.umbraContract.address, ethers.constants.MaxUint256);
+        await dai.connect(signer).approve(umbra.batchSendContract.address, ethers.constants.MaxUint256);
+      };
+
+      it('reverts if sender does not have enough tokens', async () => {
+        const msg = `Insufficient balance to complete transfer. Has 0 tokens, tried to send ${quantity.toString()} tokens.`;
+
+        if (test.id === 'send') {
+          await expectRejection(umbra.send(sender, dai.address, quantity, receiver.address), msg);
+        } else if (test.id === 'batchSend') {
+          await expectRejection(
+            umbra.batchSend(
+              sender,
+              [{ token: dai.address, amount: quantity, address: receiver!.publicKey }],
+              overrides
+            ),
+            msg
+          );
+        }
+      });
+
+      it('reverts if sender does not have enough ETH', async () => {
+        // ETH balance is checked by ethers when sending a transaction and therefore does not need to
+        // be tested here. If the user has insufficient balance it will throw with
+        // `insufficient funds for gas * price + value`
+      });
+
+      it('Send tokens, scan for them, withdraw them (direct withdraw)', async () => {
+        // SENDER
+        // Mint Dai to sender, and approve the Umbra contract to spend their DAI
+        await mintAndApproveDai(sender, sender.address, quantity.mul(5));
+
+        // Send funds with Umbra
+        let tx: ContractTransaction | null = null;
+        let stealthKeyPairs: KeyPair[] = [];
+        let usedReceivers: Wallet[] = [];
+
+        if (test.id === 'send') {
+          const result = await umbra.send(sender, dai.address, quantity, receiver!.publicKey, overrides);
+          tx = result.tx;
+          stealthKeyPairs = [result.stealthKeyPair];
+          usedReceivers = receivers.slice(0, 1);
+        } else if (test.id === 'batchSend') {
+          const sends: SendBatch[] = [];
+          for (let i = 0; i < 5; i++) {
+            sends.push({ token: dai.address, amount: quantity, address: receivers[i].publicKey });
+          }
```

Comment:
Can we get some more entropy in here? Maybe make the quantity a function of `i` or something (just ensure it's never zero). Would be nice to mix up the tokens too if that's doable. You can imagine a bug in Umbra.ts where we accidentally overwrite data with a different array item's data, and this would not catch that since all entries are the same

Reply:
Also noticed our `Send ETH, scan for it, withdraw it (direct withdraw)` is exactly the same as `Send ETH, scan for it, withdraw it (relayer withdraw)`. This is because we can't relay eth withdrawals I'm guessing? So should I remove the relayer withdraw test for eth?

Reply:
> So should I remove the relayer withdraw test for eth?

Let's track in a separate issue, this PR is already pretty big and has been open for a while, so I don't want to increase it's scope further

Reply:
Opened an issue https://github.com/ScopeLift/umbra-protocol/issues/485